src/Npgsql.SourceGenerators/DatabaseInfoSourceGenerator.cs

```diff
@@ -0,0 +1,643 @@
+            {
+                switch (c)
+                {
+                case '\r':
+                    lineNo++;
+                    posInLine = 0;
+                    outerState = ParseState.AfterArray;
+                    state = ParseState.Cr;
+                    break;
+                case '\n':
+                    lineNo++;
+                    posInLine = 0;
+                    break;
+                case '#':
+                    outerState = ParseState.AfterArray;
+                    state = ParseState.Comment;
+                    break;
+                default:
+                {
+                    if (char.IsWhiteSpace(c))
+                    {
+                        break;
+                    }
+
+                    throw new DatFileFormatException(i, lineNo, posInLine);
+                }
+                }
+                break;
+            }
+            case ParseState.Comment:
+            {
+                switch (c)
+                {
+                case '\r':
+                    lineNo++;
+                    posInLine = 0;
+                    state = ParseState.Cr;
+                    break;
+                case '\n':
+                    lineNo++;
+                    posInLine = 0;
+                    state = outerState;
+                    break;
+                }
+                break;
+            }
+            case ParseState.Cr:
+            {
+                if (c == '\n')
+                {
+                    state = outerState;
+                    break;
+                }
+
+                // undo reading this character
+                // don't increment line position
+                i--;
+                continue;
+            }
+            // ReSharper disable once UnreachableSwitchCaseDueToIntegerAnalysis
+            default:
+            {
+                throw new ArgumentOutOfRangeException();
+            }
+
+            static bool IsValidKeyChar(char c)
+                => c == '_' || char.IsLetter(c);
+            }
+
+            posInLine++;
+        }
+            
+        return list;
+    }
+
+    enum ParseState
+    {
+        Initial,
+        Comment,
+        Cr,
+        InArray,
+        InHash,
+        Key,
+        AfterKey,
+        FatComma,
+        AfterFatComma,
+        Value,
+        ValueEscape,
+        AfterValue,
+        AfterHash,
+        AfterArray,
+    }
+
+    public sealed class DatFileFormatException : FormatException
+    {
+        public int Position { get; }
+        public int LineNumber { get; }
+        public int LinePosition { get; }
+
+        public DatFileFormatException(int position, int lineNumber, int linePosition) : base(
+            $"Syntax error in dat file at line {lineNumber}, position {linePosition}")
+        {
+            Position = position;
+            LineNumber = lineNumber;
+            LinePosition = linePosition;
+        }
+    }
+
+    sealed class DatFileObject
```

Comment:
Could we just use `Dictionary<string, string>` instead of having a type?

Reply:
Yes. That's a leftover from some over ambitioned experiments with dynamic that didn't lead to a good result.